### PR TITLE
[Release/Horovod] Add user test for Horovod

### PR DESCRIPTION
**Pull request by amogkam**
Adds a "realistic" test for Horovod on Ray. The purpose of this test is to test the library in a particular environment setting that we care about.

This PR adds a new test suite for these realistic tests. All tests in this test suite must:
1. Be run in a distributed (multi-node) fashion
2. Use autoscaling/scale up (no wait_cluster.py)
3. Use GPUs if applicable
4. Have the `use_connect` flag set.

Tests run in this suite should also be located inside `ray/python/ray` so that they can be imported to other repos if needed.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by krfricke — APPROVED** [review]
LGTM do we have a nightly test output?

**Review by krfricke — DISMISSED** [review]
Ah there seems to be an error:

https://buildkite.com/ray-project/periodic-ci/builds/1336#bcc1019f-d594-45dc-9e19-b65fa51d222c

**Comment by amogkam** [comment]
Successful run: https://buildkite.com/ray-project/periodic-ci/builds/1421#b8935cf2-a733-43f2-922f-727acbfb76d4

File: release/horovod_tests/workloads/horovod_user_test.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        use_cuda=True,
+        num_epochs=20,
+        placement_group_timeout_s=900)
+
+    taken = time.time() - start
+    result = {
+        "time_taken": taken,
+    }
+    test_output_json = os.environ.get("TEST_OUTPUT_JSON",
+                                      "/tmp/horovod_releastic_test.json")
```
Comment (amogkam) [inline release/horovod_tests/workloads/horovod_user_test.py:30]:
```suggestion
                                      "/tmp/horovod_user_test.json")
```

**Comment by amogkam** [comment]
Final successful build: https://buildkite.com/ray-project/periodic-ci/builds/1460#fbc31d12-fc4d-4fd5-addb-1f6b2bb2b47f

File: python/ray/util/horovod/horovod_example.py [hunk 2]
```diff
@@ -115,11 +115,20 @@ def train_fn(data_dir=None,
...
 
 
-def main(num_workers, use_gpu, **kwargs):
-    settings = RayExecutor.create_settings(timeout_s=30)
+def main(num_workers,
+         use_gpu,
+         timeout_s=30,
+         placement_group_timeout_s=100,
+         kwargs=None):
+    kwargs = kwargs if kwargs else {}
```
Comment (richardliaw) [inline python/ray/util/horovod/horovod_example.py:123]: nit, kwargs = kwargs or {}

**Review by richardliaw — CHANGES_REQUESTED** [review]

**Review by richardliaw — APPROVED** [review]

File: release/.buildkite/build_pipeline.py [hunk 3]
```diff
@@ -474,22 +496,21 @@ def create_test_step(
...
-    ]
-
     if isinstance(test_name, ConnectTest):
         # Add driver side setup commands to the step.
         pip_requirements_command = [f"pip install -U -r "
                                     f"{test_name.requirements_file}"] if \
             test_name.requirements_file else []
         step_conf["commands"] = test_name.setup_commands \
-            + pip_requirements_command \
-            + step_conf["commands"]
```
[the remark this answers is not in the document]
Comment (amogkam) [reply release/.buildkite/build_pipeline.py:492]: We want to do the necessary driver side setup before running the e2e script.

File: release/horovod_tests/driver_requirements.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+# The cluster uses ray-ml Docker image.
+# ray-ml Docker image installs dependencies from ray/python/requirements/ml/ directory.
+# We constrain on these requirements file so that the same versions are installed.
+-c ../../python/requirements/ml/requirements_dl.txt
+
+torch
+torchvision
+
+# Horovod master.
+git+https://github.com/horovod/horovod.git
```
[the remark this answers is not in the document]
Comment (amogkam) [reply release/horovod_tests/driver_requirements.txt:11]: that could lead to problems where the error is not uncovered until after the release is made, right?

[the remark this answers is not in the document]
Comment (amogkam) [reply release/horovod_tests/driver_requirements.txt:11]: I think we might want to test both.

**Pull request merged**